File: Android/sqo/src/main/java/dotissi/sqo/QueryRunner.java
package dotissi.sqo;

import org.fusesource.lmdbjni.Transaction;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;


public class QueryRunner {

    TagsIndexManager indexManag;
    Transaction lmdbTransaction;
    String BucketName;
    TransactionManager transactionManager;
    public QueryRunner(TagsIndexManager indexManag, Transaction lmdbTransaction, String bucketName,TransactionManager transactionManager) {
        this.indexManag = indexManag;
        this.lmdbTransaction = lmdbTransaction;
        this.BucketName = bucketName;
        this.transactionManager=transactionManager;
    }

    public ArrayList<String> run(Query query) {
        if (query.wheres.size() == 0)
        {
            throw new IllegalArgumentException("Query does not have defined any filtering");
        }
        ArrayList<Where> uniqueWheres = new ArrayList<Where>();
        uniqueWheres.addAll(query.wheres);
        this.TryOptimizeBetween(query, uniqueWheres);
        this.MakeSorting(query, uniqueWheres);

        //intersection
        ArrayList<String> keys1 = GetBySingleWhere(uniqueWheres.get(0));
        for (int i = 1; i < uniqueWheres.size(); i++)
        {
            if (keys1.size() == 0)
                break;
            ArrayList<String>  keys2 = GetBySingleWhere(uniqueWheres.get(i));
            keys1.retainAll(keys2);
        }
        //TODO optimize this
        //union
        for(Query or:query.ors)
        {
            ArrayList<String> keys2 = this.run(or);
            HashSet<String> set=new HashSet<String>();
            set.addAll(keys1);

            for(String key: keys2)
            {
               set.add(key);
            }
            keys1=new ArrayList<String>();
            keys1.addAll(set);
        }
        return keys1;

    }
    private void TryOptimizeBetween(Query query, ArrayList<Where> uniqueWheres) {
        HashMap<String, Where> groups = new HashMap<String, Where>();
        ArrayList<Where> tobeRemoved=new ArrayList<>();
        ArrayList<Where> tobeAdded=new ArrayList<>();

        for (Where w : uniqueWheres) {
            if (!groups.containsKey(w.TagName)) {
                groups.put(w.TagName, w);
            } else {
                Where dup = groups.get(w.TagName);
                if ((w.Operator == WhereOp.GreaterThan || w.Operator == WhereOp.GreaterThanOrEqual) &&
                        (dup.Operator == WhereOp.LessThan || dup.Operator == WhereOp.LessThanOrEqual)) {
                    tobeRemoved.add(w);
                    tobeRemoved.add(dup);
                    Where wNew = new Where(w.TagName);
                    wNew.Between = new Object[] { w.Value, dup.Value };
                    wNew.Operator = this.getBetweenOperator(w.Operator, dup.Operator);
                    tobeAdded.add(wNew);
                }
                else if ((dup.Operator == WhereOp.GreaterThan || dup.Operator == WhereOp.GreaterThanOrEqual) &&
                        (w.Operator == WhereOp.LessThan || w.Operator == WhereOp.LessThanOrEqual)) {

                    tobeRemoved.add(w);
                    tobeRemoved.add(dup);
                    Where wNew = new Where(w.TagName);
                    wNew.Between = new Object[] { dup.Value, w.Value };
                    wNew.Operator = this.getBetweenOperator(dup.Operator, w.Operator);
                    tobeAdded.add(wNew);
                }
                groups.put(w.TagName,w);

            }
        }

    }
    private void MakeSorting(Query query, ArrayList<Where> uniqueWheres)
    {
        for (int i = 0; i < query.orderby.size(); i++)
        {
            int index =this.GetIndexForTagName(uniqueWheres, query.orderby.get(i).tagName);
            if (index >= 0)
            {
                uniqueWheres.get(index).Descending = query.orderby.get(i).desc;
                if (index != i)
                {
                    Where w = uniqueWheres.get(index);
                    uniqueWheres.remove(index);
                    uniqueWheres.add(i, w);
                }

            }
        }
    }
    private int GetIndexForTagName(ArrayList<Where> wheres,String tagName)
    {
        for(int i=0;i<wheres.size();i++)
        {
            if(wheres.get(i).TagName==tagName)
                return i;
        }
        return -1;
    }
    private ArrayList<String> GetBySingleWhere(Where where) {
        if (where.TagName == "key") {
            return this.LoadByKey(where, lmdbTransaction);
        } else //by tags
        {
            return this.indexManag.LoadKeysByIndex(where, this.BucketName, lmdbTransaction);
        }
    }

    private ArrayList<String> LoadByKey(Where where, Transaction lmdbTransaction) {
        IIndex index = new IndexKey(this.BucketName, lmdbTransaction,transactionManager);
        return IndexQueryFinder.FindKeys(index, where);
    }

    private WhereOp getBetweenOperator(WhereOp startOp, WhereOp endOp)
    {
        if (startOp == WhereOp.GreaterThan && endOp == WhereOp.LessThan)
            return WhereOp.BetweenExceptStartEnd;
        else if(startOp == WhereOp.GreaterThanOrEqual && endOp == WhereOp.LessThan)
            return WhereOp.BetweenExceptEnd;
        else if (startOp == WhereOp.GreaterThan && endOp == WhereOp.LessThanOrEqual)
            return WhereOp.BetweenExceptStart;
        else if (startOp == WhereOp.GreaterThanOrEqual && endOp == WhereOp.LessThanOrEqual)
            return WhereOp.Between;

        return WhereOp.Between;
    }
}

File: siaqodb/Documents/Queries/QueryRunner.cs
using LightningDB;
using Sqo.Documents.Indexes;
using Sqo.Exceptions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


namespace Sqo.Documents.Queries
{
    class QueryRunner
    {
        TagsIndexManager indexManag;
        LightningTransaction lmdbTransaction;
        string BucketName;
        public QueryRunner(TagsIndexManager indexManag,LightningTransaction lmdbTransaction, string bucketName)
        {
            this.indexManag = indexManag;
            this.lmdbTransaction = lmdbTransaction;
            this.BucketName = bucketName;
        }
        public List<string> Run(Query query)
        {
            if (query.wheres.Count == 0)
            {
                throw new ArgumentException("Query does not have defined any filtering");
            }
            List<Where> uniqueWheres = new List<Where>();
            uniqueWheres.AddRange(query.wheres);
            this.TryOptimizeBetween(query, uniqueWheres);
            this.MakeSorting(query, uniqueWheres);

            //intersection
            var keys1 = GetBySingleWhere(uniqueWheres[0]);
            for (int i = 1; i < uniqueWheres.Count; i++)
            {
                if (keys1.Count == 0)
                    break;
                var keys2 = GetBySingleWhere(uniqueWheres[i]);
                keys1 = keys1.Intersect(keys2).ToList();
            }
            //union
            foreach(Query or in query.ors)
            {
                var keys2 = this.Run(or);
                var dict = keys1.ToDictionary(a => a);
                foreach (var key in keys2)
                {
                    dict[key] = key;
                }
                keys1 = dict.Values.ToList();
            }
            return keys1;

        }

        private void TryOptimizeBetween(Query query, List<Where> uniqueWheres)
        {
            var dup = query.wheres.GroupBy(x => x.TagName)
                  .Where(g => g.Count() > 1)
                  .Select(y => new { TagName = y.Key, Counter = y.Count() })
                  .ToList();
            foreach (var groupItem in dup)
            {
                var startWhere = query.wheres.Find(a => a.TagName == groupItem.TagName && (a.Operator == WhereOp.GreaterThan || a.Operator == WhereOp.GreaterThanOrEqual));
                var endWhere = query.wheres.Find(a => a.TagName == groupItem.TagName && (a.Operator == WhereOp.LessThan || a.Operator == WhereOp.LessThanOrEqual));
                if (startWhere != null && endWhere != null)
                {
                    Where wNew = new Where(groupItem.TagName);
                    wNew.Between = new object[] { startWhere.Value, endWhere.Value };
                    wNew.Operator = this.GetBetweenOperator(startWhere.Operator, endWhere.Operator);
                    uniqueWheres.Add(wNew);
                    uniqueWheres.Remove(startWhere);
                    uniqueWheres.Remove(endWhere);
                }
            }
        }

        private void MakeSorting(Query query, List<Where> uniqueWheres)
        {
            for (int i = 0; i < query.orderby.Count; i++)
            {
                int index = uniqueWheres.FindIndex(a => a.TagName == query.orderby[i].tagName);
                if (index >= 0)
                {
                    uniqueWheres[index].Descending = query.orderby[i].desc;
                    if (index != i)
                    {
                        Where w = uniqueWheres[index];
                        uniqueWheres.RemoveAt(index);
                        uniqueWheres.Insert(i, w);
                    }

                }
            }
        }

        private List<string> GetBySingleWhere(Where where)
        {
            if (string.Compare(where.TagName, "key", StringComparison.OrdinalIgnoreCase) == 0)
            {
                return this.LoadByKey(where, lmdbTransaction);
            }
            else //by tags
            {
                return this.indexManag.LoadKeysByIndex(where, this.BucketName, lmdbTransaction);
            }
        }

        private WhereOp GetBetweenOperator(WhereOp startOp, WhereOp endOp)
        {
            if (startOp == WhereOp.GreaterThan && endOp == WhereOp.LessThan)
                return WhereOp.BetweenExceptStartEnd;
            else if(startOp == WhereOp.GreaterThanOrEqual && endOp == WhereOp.LessThan)
                return WhereOp.BetweenExceptEnd;
            else if (startOp == WhereOp.GreaterThan && endOp == WhereOp.LessThanOrEqual)
                return WhereOp.BetweenExceptStart;
            else if (startOp == WhereOp.GreaterThanOrEqual && endOp == WhereOp.LessThanOrEqual)
                return WhereOp.Between;

            return WhereOp.Between;
        }

        private List<string> LoadByKey(Where where, LightningTransaction transaction)
        {
            var db = transaction.OpenDatabase(BucketName, DatabaseOpenFlags.Create);
            {
                IIndex index = new IndexKey(db, transaction);
                return IndexQueryFinder.FindKeys(index, where);
            }
        }
    }
}
